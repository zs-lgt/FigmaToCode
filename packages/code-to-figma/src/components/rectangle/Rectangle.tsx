import * as React from 'react';
import {
    DefaultShapeProps,
    BorderProps,
    CornerProps,
    StyleOf,
    InstanceItemProps,
    SelectionEventProps
} from '../../types';
import {
    LayoutStyleProperties,
    transformLayoutStyleProperties
} from '../../styleTransformers/transformLayoutStyleProperties';
import {
    GeometryStyleProperties,
    transformGeometryStyleProperties
} from '../../styleTransformers/transformGeometryStyleProperties';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import {
    BorderStyleProperties,
    transformBorderStyleProperties
} from '../../styleTransformers/transformBorderProperties';
import { transformBlendProperties, BlendStyleProperties } from '../../styleTransformers/transformBlendProperties';
import { StyleSheet } from '../..';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { useSelectionChange } from '../../hooks/useSelectionChange';
import { transformAutoLayoutToYoga } from '../../styleTransformers/transformAutoLayoutToYoga';
import { OnLayoutHandlerProps, useOnLayoutHandler } from '../../hooks/useOnLayoutHandler';
import { filter } from 'rxjs/operators';
import { useImageHash } from '../../hooks/useImageHash';
import { useNodeIdCallback } from '../../hooks/useNodeIdCallback';

export interface RectangleProps
    extends DefaultShapeProps,
        CornerProps,
        BorderProps,
        InstanceItemProps,
        SelectionEventProps,
        OnLayoutHandlerProps {
    style?: StyleOf<
        LayoutStyleProperties &
            YogaStyleProperties &
            BorderStyleProperties &
            BlendStyleProperties &
            GeometryStyleProperties
    >;
    children?: undefined;
}

const Rectangle: React.FC<RectangleProps> = props => {
    const nodeRef = React.useRef();

    useSelectionChange(nodeRef, props);
    useNodeIdCallback(nodeRef, props.onNodeId);

    const style = { ...StyleSheet.flatten(props.style), ...transformAutoLayoutToYoga(props) };

    const imageHash = useImageHash(style.backgroundImage);

    const rectangleProps = {
        ...transformLayoutStyleProperties(style),
        ...transformGeometryStyleProperties('fills', style, imageHash),
        ...transformBorderStyleProperties(style),
        ...transformBlendProperties(style),
        ...props,
        style
    };
    const yogaProps = useYogaLayout({ nodeRef, ...rectangleProps });

    useOnLayoutHandler(yogaProps, props);

    return <rectangle {...rectangleProps} {...yogaProps} innerRef={nodeRef} />;
};

export { Rectangle };
