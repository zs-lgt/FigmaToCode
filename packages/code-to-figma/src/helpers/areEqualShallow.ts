export const areEqualShallow = (obj1?: Object, obj2?: Object): boolean => {
    if (Object.is(obj1, obj2)) {
        return true;
    }

    if (typeof obj1 !== 'object' || obj1 === null || typeof obj2 !== 'object' || obj2 === null) {
        return false;
    }

    return (
        Object.keys(obj1).length === Object.keys(obj2).length &&
        Object.keys(obj1).every(key => obj2.hasOwnProperty(key) && obj1[key] === obj2[key])
    );
};
