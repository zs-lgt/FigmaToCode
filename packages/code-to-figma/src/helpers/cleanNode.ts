export const cleanNode = node => {
    if (!node) {
        return;
    }
    const { children, parent, ref, innerRef, ...other } = node;
    return other;
};
