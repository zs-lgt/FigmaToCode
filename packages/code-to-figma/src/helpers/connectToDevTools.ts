import { connectToDevTools as connectToDevToolsInternal } from 'react-devtools-core';
import { api } from '../rpc';

export const connectToDevTools = () => {
    let connectionFailed = false;
    const connection = new WebSocket('ws://localhost:8097');

    connectToDevToolsInternal({
        // prevents from multiple retries
        isAppActive: () => !connectionFailed,
        resolveRNStyle: null,
        websocket: connection
    });

    connection.onerror = () => {
        console.warn('React DevTools is not running. Please run React DevTools or remove `connectToDevTools` call');
        connectionFailed = true;
    };

    // @ts-ignore
    window.__REACT_DEVTOOLS_GLOBAL_HOOK__.on('react-devtools', agent => {
        agent._bridge.addListener('inspectElement', data => {
            const { rendererID, id } = data;

            const renderer = agent.rendererInterfaces[rendererID];

            let nodes = null;
            if (renderer !== null) {
                nodes = renderer.findNativeNodesForFiberID(id);
            }

            if (nodes !== null && nodes[0]) {
                api.highlightNativeElement(nodes[0]);
            }
        });
    });
};
