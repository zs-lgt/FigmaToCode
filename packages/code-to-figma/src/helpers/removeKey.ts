export const removeKey = node => {
    if (node && node.key) {
        delete node.key;
    }
    if (node && node.children) {
        node.children.forEach(removeKey);
    }
    return node;
};
