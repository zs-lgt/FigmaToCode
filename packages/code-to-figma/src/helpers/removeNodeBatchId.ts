export const removeNodeBatchId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.nodeBatchId;
    }
    if (node && node.children) {
        node.children.forEach(removeNodeBatchId);
    }
    return node;
};
