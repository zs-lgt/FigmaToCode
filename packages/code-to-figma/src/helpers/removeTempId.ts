export const removeTempId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.reactId;
    }
    if (node && node.children) {
        node.children.forEach(removeTempId);
    }
    return node;
};
