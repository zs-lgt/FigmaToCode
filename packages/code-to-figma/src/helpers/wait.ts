import { act } from 'react-test-renderer';

export function wait(amount = 0) {
    return new Promise(resolve => setTimeout(resolve, amount));
}

// Use this in your test after mounting if you need just need to let the query finish without updating the wrapper
export async function actWait(amount = 0) {
    await act(async () => {
        await wait(amount);
    });
}
