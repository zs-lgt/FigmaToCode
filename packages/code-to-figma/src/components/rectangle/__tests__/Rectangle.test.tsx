import * as React from 'react';
const renderer = require('react-test-renderer');
import { Rectangle } from '../Rectangle';
import { createFigma } from 'figma-api-stub';
import { wait } from '../../../helpers/wait';
import { render } from '../../../renderer';
import { removeNodeBatchId } from '../../../helpers/removeNodeBatchId';
import { removeTempId } from '../../../helpers/removeTempId';
import { Page } from '../../..';
import { removeKey } from '../../../helpers/removeKey';

const removeMeta = node => {
    return removeKey(removeNodeBatchId(removeTempId(node)));
};

describe('<Rectangle />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('Rectangle without props', () => {
        const tree = renderer.create(<Rectangle />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with layout style properties', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        width: 100,
                        height: 200,
                        top: 50,
                        left: 20
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with color background', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        backgroundColor: '#1bff00'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with image background', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        backgroundColor: '#ff8b29',
                        backgroundSize: 'cover',
                        backgroundImage: './image.png'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with border props', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        width: 200,
                        height: 100,
                        borderWidth: 10,
                        borderColor: '#ff8b29'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Loading backgroundImage', async () => {
        // @ts-ignore
        global.figma.createImage = () => {
            return {
                hash: '<image hash>'
            };
        };

        // @ts-ignore
        global.FileReader = () => {
            const obj: any = {
                readAsDataURL: () => {},
                readAsArrayBuffer: () => {}
            };

            setTimeout(() => {
                obj.result = '<test hash>';
                if (obj.onloadend) {
                    obj.onloadend();
                }
                if (obj.onload) {
                    obj.onload();
                }
            });

            return obj;
        };

        await render(
            <Page>
                <Rectangle
                    style={{
                        backgroundImage: 'https://avatars1.githubusercontent.com/u/54585419'
                    }}
                />
            </Page>
        );
        await wait();
        await wait();
        await wait();
        await wait();
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });
});
