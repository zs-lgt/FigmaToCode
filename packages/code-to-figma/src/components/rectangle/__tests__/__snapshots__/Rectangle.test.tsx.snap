// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rectangle /> Loading backgroundImage 1`] = `
ChildrenMixinStub {
  "children": Array [
    ExportMixinStub {
      "children": Array [],
      "config": undefined,
      "id": "0:1",
      "parent": [Circular],
      "type": "PAGE",
    },
    ExportMixinStub {
      "_backgrounds": Array [
        Object {
          "blendMode": "NORMAL",
          "color": Object {
            "b": 0.9607843160629272,
            "g": 0.9607843160629272,
            "r": 0.9607843160629272,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "children": Array [
        GeometryMixinStub {
          "_fills": Array [
            Object {
              "imageHash": "<image hash>",
              "scaleMode": "FILL",
              "type": "IMAGE",
            },
          ],
          "blendMode": "NORMAL",
          "config": Object {
            "isWithoutTimeout": false,
            "simulateErrors": true,
          },
          "constraints": Object {
            "horizontal": "MIN",
            "vertical": "MIN",
          },
          "cornerRadius": 0,
          "cornerSmoothing": 0,
          "effects": Array [],
          "exportSettings": Array [],
          "id": "1:2",
          "isMask": false,
          "layoutAlign": "INHERIT",
          "layoutGrow": 0,
          "locked": false,
          "opacity": 1,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"backgroundImage\\":\\"https://avatars1.githubusercontent.com/u/54585419\\"}",
          },
          "strokeAlign": "INSIDE",
          "strokeCap": "NONE",
          "strokeWeight": 0,
          "strokes": Array [],
          "type": "RECTANGLE",
          "visible": true,
          "x": 0,
          "y": 0,
        },
      ],
      "config": Object {
        "isWithoutTimeout": false,
        "simulateErrors": true,
      },
      "exportSettings": Array [],
      "id": "1:1",
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "config": undefined,
  "id": "0:0",
  "type": "DOCUMENT",
}
`;

exports[`<Rectangle /> Rectangle with border props 1`] = `
<rectangle
  height={100}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  strokeWeight={10}
  strokes={
    Array [
      Object {
        "color": Object {
          "b": 0.1607843137254902,
          "g": 0.5450980392156862,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  style={
    Object {
      "borderColor": "#ff8b29",
      "borderWidth": 10,
      "height": 100,
      "width": 200,
    }
  }
  width={200}
/>
`;

exports[`<Rectangle /> Rectangle with color background 1`] = `
<rectangle
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 1,
          "r": 0.10588235294117647,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#1bff00",
    }
  }
/>
`;

exports[`<Rectangle /> Rectangle with image background 1`] = `
<rectangle
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0.1607843137254902,
          "g": 0.5450980392156862,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#ff8b29",
      "backgroundImage": "./image.png",
      "backgroundSize": "cover",
    }
  }
/>
`;

exports[`<Rectangle /> Rectangle with layout style properties 1`] = `
<rectangle
  height={200}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "height": 200,
      "left": 20,
      "top": 50,
      "width": 100,
    }
  }
  width={100}
  x={20}
  y={50}
/>
`;

exports[`<Rectangle /> Rectangle without props 1`] = `
<rectangle
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={Object {}}
/>
`;
